File: src/domain/validators.py
"""Domain validation logic for network records."""

import re
from functools import lru_cache
from typing import Final, Protocol

from pydantic import BaseModel, IPvAnyAddress, IPvAnyNetwork, ValidationError

from .entities import NetworkRecord, RecordType


class Validator(Protocol):
    """Protocol for validators."""
    
    def is_valid(self, value: str) -> bool:
        """Check if value is valid."""
        ...


class FQDNValidator:
    """
    Validates Fully Qualified Domain Names per RFC 1035 and RFC 3696.
    
    Uses caching for improved performance on repeated validations.
    """
    
    MAX_LENGTH: Final[int] = 253
    MIN_LABELS: Final[int] = 2
    
    _PATTERN: Final[re.Pattern] = re.compile(
        r"^((?![-])[-A-Z\d]{1,63}(?<!-)[.])*(?!-)[-A-Z\d]{1,63}(?<!-)[.]?$",
        re.IGNORECASE
    )
    
    @classmethod
    @lru_cache(maxsize=1024)
    def is_valid(cls, fqdn: str) -> bool:
        """
        Validate FQDN with caching.
        
        Args:
            fqdn: Domain name to validate
            
        Returns:
            True if valid FQDN
        """
        if not isinstance(fqdn, str) or not fqdn:
            return False
        
        # Handle wildcard domains
        normalized = fqdn.removeprefix("*.").lower()
        
        # Check length
        if len(normalized.rstrip(".")) > cls.MAX_LENGTH:
            return False
        
        # Check pattern
        if not cls._PATTERN.match(normalized):
            return False
        
        # Check minimum labels
        label_count = normalized.rstrip(".").count(".") + 1
        return label_count >= cls.MIN_LABELS


class IPAddressValidator:
    """Validates IPv4 and IPv6 addresses using Pydantic."""
    
    @staticmethod
    @lru_cache(maxsize=1024)
    def is_valid(ip: str) -> bool:
        """
        Validate IP address with caching.
        
        Args:
            ip: IP address string
            
        Returns:
            True if valid IPv4 or IPv6
        """
        if not isinstance(ip, str) or not ip:
            return False
        
        class _IPModel(BaseModel):
            address: IPvAnyAddress
        
        try:
            _IPModel(address=ip)
            return True
        except ValidationError:
            return False


class NetworkCIDRValidator:
    """Validates network addresses in CIDR notation."""
    
    @staticmethod
    @lru_cache(maxsize=1024)
    def is_valid(network: str) -> bool:
        """
        Validate network CIDR with caching.
        
        Args:
            network: Network in CIDR notation
            
        Returns:
            True if valid CIDR
        """
        if not isinstance(network, str) or "/" not in network:
            return False
        
        class _NetworkModel(BaseModel):
            network: IPvAnyNetwork
        
        try:
            _NetworkModel(network=network)
            return True
        except ValidationError:
            return False


class RecordClassifier:
    """
    Classifies network records into their appropriate types.
    
    Uses the strategy pattern with multiple validators.
    """
    
    def __init__(self):
        """Initialize classifier with validators."""
        self._validators = [
            (NetworkCIDRValidator, RecordType.NETWORK_CIDR),
            (IPAddressValidator, RecordType.IP_ADDRESS),
            (FQDNValidator, RecordType.FQDN),
        ]
    
    def classify(self, value: str) -> NetworkRecord:
        """
        Classify a network record string.
        
        Args:
            value: Raw network record string
            
        Returns:
            NetworkRecord with determined type
        """
        value = value.strip()
        
        for validator, record_type in self._validators:
            if validator.is_valid(value):
                return NetworkRecord(value=value, record_type=record_type)
        
        return NetworkRecord(value=value, record_type=RecordType.INVALID)
    
    def classify_batch(self, values: list[str]) -> list[NetworkRecord]:
        """
        Classify multiple records.
        
        Args:
            values: List of raw record strings
            
        Returns:
            List of classified NetworkRecords
        """
        return [self.classify(value) for value in values]


# Convenience functions for backward compatibility
def is_valid_fqdn(fqdn: str) -> bool:
    """Check if string is a valid FQDN."""
    return FQDNValidator.is_valid(fqdn)


def is_valid_ip_address(ip: str) -> bool:
    """Check if string is a valid IP address."""
    return IPAddressValidator.is_valid(ip)


def is_valid_network_cidr(network: str) -> bool:
    """Check if string is valid network CIDR."""
    return NetworkCIDRValidator.is_valid(network)

File: src/rust-app/src/domain/validators.rs
//! Domain validation logic for network records. Ports `src/domain/validators.py`.
//!
//! The Python validators delegate to pydantic (`IPvAnyAddress`/`IPvAnyNetwork`)
//! and a regex. Those semantics are reproduced here without a regex engine:
//!
//! * IPv4/IPv6 parsing uses `std::net`, which matches Python `ipaddress`:
//!   leading-zero octets are rejected, surrounding whitespace is rejected.
//!   Python additionally accepts an IPv6 `%zone` scope id, so that suffix is
//!   stripped before parsing an address.
//! * Networks are validated *strictly* (host bits must be zero), matching
//!   pydantic `IPvAnyNetwork`. The prefix must be all decimal digits.
//! * FQDNs are validated label-by-label, matching the Python regex.

use std::net::{Ipv4Addr, Ipv6Addr};

use ipnet::{IpNet, Ipv4Net, Ipv6Net};

use super::entities::{NetworkRecord, RecordType};

const FQDN_MAX_LENGTH: usize = 253;
const FQDN_MIN_LABELS: usize = 2;

// --- FQDN -------------------------------------------------------------------

/// Validate a Fully Qualified Domain Name (RFC 1035 / 3696).
pub fn fqdn_is_valid(fqdn: &str) -> bool {
    if fqdn.is_empty() {
        return false;
    }

    // Handle wildcard domains: Python `removeprefix("*.")` strips one prefix.
    let normalized = match fqdn.strip_prefix("*.") {
        Some(rest) => rest.to_lowercase(),
        None => fqdn.to_lowercase(),
    };

    // Length check uses the value with all trailing dots removed.
    let trimmed = normalized.trim_end_matches('.');
    if trimmed.chars().count() > FQDN_MAX_LENGTH {
        return false;
    }

    if !fqdn_pattern_matches(&normalized) {
        return false;
    }

    // Minimum label count.
    let label_count = trimmed.matches('.').count() + 1;
    label_count >= FQDN_MIN_LABELS
}

/// Reproduces the Python FQDN regex: zero or more `label.` groups followed by
/// a final `label` and an optional single trailing dot.
fn fqdn_pattern_matches(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    // The regex permits exactly one optional trailing dot.
    let core = s.strip_suffix('.').unwrap_or(s);
    if core.is_empty() {
        return false;
    }
    core.split('.').all(label_ok)
}

/// A single DNS label: 1-63 ASCII alphanumerics/hyphens, no leading/trailing hyphen.
fn label_ok(label: &str) -> bool {
    if label.is_empty() {
        return false;
    }
    if !label.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'-') {
        return false;
    }
    // All bytes are ASCII here, so byte length equals character length.
    if label.len() > 63 {
        return false;
    }
    !label.starts_with('-') && !label.ends_with('-')
}

// --- IP address -------------------------------------------------------------

/// Validate an IPv4 or IPv6 address.
pub fn ip_is_valid(ip: &str) -> bool {
    if ip.is_empty() {
        return false;
    }
    ip.parse::<Ipv4Addr>().is_ok() || ipv6_is_valid(ip)
}

/// IPv6 parsing that, like Python `ipaddress`, accepts an optional `%zone`.
fn ipv6_is_valid(ip: &str) -> bool {
    let base = match ip.split_once('%') {
        Some((_, zone)) if zone.is_empty() => return false,
        Some((addr, _)) => addr,
        None => ip,
    };
    base.parse::<Ipv6Addr>().is_ok()
}

// --- Network CIDR -----------------------------------------------------------

/// Validate a network in CIDR notation (strict: host bits must be zero).
pub fn cidr_is_valid(network: &str) -> bool {
    if !network.contains('/') {
        return false;
    }
    match parse_network_lax(network) {
        Some(net) => net.trunc() == net,
        None => false,
    }
}

/// Parse `addr/prefix` permissively (host bits allowed). Returns `None` for any
/// malformed address or prefix. The prefix must be all decimal digits, matching
/// Python `ipaddress` (which rejects `+`, signs and surrounding whitespace but
/// tolerates leading zeros).
fn parse_network_lax(network: &str) -> Option<IpNet> {
    let (addr_s, prefix_s) = network.split_once('/')?;
    if prefix_s.is_empty() || !prefix_s.bytes().all(|b| b.is_ascii_digit()) {
        return None;
    }
    let prefix: u32 = prefix_s.parse().ok()?;

    if let Ok(addr) = addr_s.parse::<Ipv4Addr>() {
        if prefix > 32 {
            return None;
        }
        return Some(IpNet::V4(Ipv4Net::new(addr, prefix as u8).ok()?));
    }
    if let Ok(addr) = addr_s.parse::<Ipv6Addr>() {
        if prefix > 128 {
            return None;
        }
        return Some(IpNet::V6(Ipv6Net::new(addr, prefix as u8).ok()?));
    }
    None
}

/// Normalize a CIDR string the way Python `str(ip_network(x, strict=False))`
/// does: host bits masked off, IPv6 lower-cased and compressed.
pub fn normalize_cidr(network: &str) -> Option<String> {
    parse_network_lax(network).map(|net| net.trunc().to_string())
}

/// Network address and dotted/compressed netmask for a CIDR string, matching
/// Python `ip_network(...).network_address` and `.netmask`.
pub fn network_parts(network: &str) -> Option<(String, String)> {
    let net = parse_network_lax(network)?;
    Some((net.network().to_string(), net.netmask().to_string()))
}

// --- Classification ---------------------------------------------------------

/// Classify a raw record string into a `NetworkRecord`.
///
/// Validators run in priority order: CIDR, then IP address, then FQDN.
pub fn classify(value: &str) -> NetworkRecord {
    let value = value.trim();
    if cidr_is_valid(value) {
        NetworkRecord::new(value, RecordType::NetworkCidr)
    } else if ip_is_valid(value) {
        NetworkRecord::new(value, RecordType::IpAddress)
    } else if fqdn_is_valid(value) {
        NetworkRecord::new(value, RecordType::Fqdn)
    } else {
        NetworkRecord::new(value, RecordType::Invalid)
    }
}

/// Classify many records.
pub fn classify_batch<I, S>(values: I) -> Vec<NetworkRecord>
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    values.into_iter().map(|v| classify(v.as_ref())).collect()
}
